doc/cephfs/mdcache.rst

```diff
@@ -0,0 +1,80 @@
+on its behalf, but an MDS can also grant the client **capabilities**
+(aka **caps**) for each inode (see :doc:`/cephfs/capabilities`).
+
+A capability grants the client the ability to cache and possibly
+manipulate some portion of the data or metadata associated with the
+inode. When another client needs access to the same information, the MDS
+will revoke the capability and the client will eventually return it,
+along with an updated version of the inode's metadata (in the event that
+it made changes to it while it held the capability).
+
+Clients can request capabilities and will generally get them, but when
+there is competing access or memory pressure on the MDS, they may be
+**revoked**. When a capability is revoked, the client is responsible for
+returning it as soon as it is able. Clients that fail to do so in a
+timely fashion may end up **blacklisted** and unable to communicate with
+the cluster.
+
+Since the cache is distributed, the MDS must take great care to ensure
+that no client holds capabilities that may conflict with other clients'
+capabilities, or operations that it does itself.
+
+This allows cephfs clients to rely on much greater cache coherence than
+a filesystem like NFS, where the client may cache data and metadata
+beyond the point where it has changed on the server.
+
+Client Metadata Requests
+------------------------
+When a client needs to query/change inode metadata or perform an
+operation on a directory, it has two options. It can make a request to
+the MDS directly, or serve the information out of its cache. With
+CephFS, the latter is only possible if the client has the necessary
+caps.
+
+Clients can send simple requests to the MDS to query or request changes
+to certain metadata. The replies to these requests may also grant the
+client a certain set of caps for the inode, allowing it to perform
+subsequent requests without consulting the MDS.
+
+Clients can also request caps directly from the MDS, which is necessary
+in order to read or write file data.
+
+MDS Inode Locker
+----------------
+When an MDS wants to read or change information about an inode, it must
+gather the appropriate locks for it. If there are outstanding caps that
+would conflict with these locks, then they must be revoked before the
+lock can be acquired. Once the competing caps are returned to the MDS,
+then it can get the locks and do the operation.
+
+Multiple MDS'
+-------------
+On a filesystem served by multiple MDS', the metadata cache is also
+distributed among the MDS' in the cluster. For every inode, at any given
+time, only one MDS in the cluster is considered **authoritative**. Any
+requests to change that inode must be done by the authoritative MDS,
+though non-authoritative MDS can forward requests to the authoritative
+one.
+
+Non-auth MDS' can also obtain read locks that prevent the auth MDS from
+changing the data until the lock is dropped, so that they can serve
+inode info to the clients. [1][2]
```

Comment:
I'd prefer to just merge this discussion of locks with the previous section and re-title it as "Distributed Locks in an MDS Cluster" or similar. Eventually we want to get to a point where we don't belabor the difference in behavior between 1 active MDS vs. N active MDS. 1 active MDS should just be a special case in the code (and it's not really).